controller/lib/core/blower_fsm.h

```diff
@@ -155,11 +155,109 @@ class PressureControlFsm {
+      // will approximately be 0 and we may assume to be back to the first case.
+      inspiratory_effort_threshold_ =
+          ml_per_min(readings.flow_ml_per_min + 12000.0f);
+      // Note 12000 ml/min looks like a lot but it is actually 200 ml/sec which
+      // is in the high end of our sensor's sensitivity for relatively low flows
+      // that are expected during exhale, especially when summing two sensors.
+      // I can see two ways to improve this sensitivity:
+      //   a: provide a denoised flow in SensorReadings, at the cost of reducing
+      //      response time
+      //   b: provide separate inflow and outflow, at the cost of a more
+      //      complicated (and not yet determined) breath detection algorithm.
+    }
+
+    // Only detect the first time that inspiratory effort is applied in order to
+    // update inspire_end_ on the first detection only.
+    if (!patient_inspiring_ && inspiratory_effort_threshold_ > ml_per_min(0) &&
+        readings.flow_ml_per_min > inspiratory_effort_threshold_.ml_per_min()) {
+      patient_inspiring_ = true;
+      // update inspire_end_ to enforce proper inspire duration even if patient
+      // started inspiring before latest_expire_end_
+      inspire_end_ -= latest_expire_end_ - now;
+    }
+
+    if (!patient_inspiring_ && now < latest_expire_end_) {
+      return {.blower_enabled = true, expire_pressure_, ValveState::OPEN};
+    }
+    // update patient inspiring so we don't attempt detecting inspiratory effort
+    // once the inspiration phase has started
+    patient_inspiring_ = true;
+    return {.blower_enabled = true, inspire_pressure_, ValveState::CLOSED};
+  }
+
+  bool finished(Time now) { return now > inspire_end_; }
+
+private:
+  // Given t = secs_per_breath and r = I:E ratio, calculate inspiration duration
+  // and maximum expiration durations (I and E).
+  static Duration inspire_duration(const VentParams &params) {
+    float t =
+        60.f / static_cast<float>(params.breaths_per_min); // secs per breath
+    float r = params.inspiratory_expiratory_ratio;         // I:E
+    return seconds(t * r / (1 + r));
+  }
+  static Duration max_expire_duration(const VentParams &params) {
+    float t =
+        60.f / static_cast<float>(params.breaths_per_min); // secs per breath
+    float r = params.inspiratory_expiratory_ratio;         // I:E
+    return seconds(t / (1 + r));
+  }
+
+  const Pressure inspire_pressure_;
+  const Pressure expire_pressure_;
+
+  Time start_time_;
+  Time latest_expire_end_;
+  Time inspire_end_;
+
+  VolumetricFlow inspiratory_effort_threshold_ = ml_per_min(-1.0f);
+  bool patient_inspiring_ = false;
+};
+
 class BlowerFsm {
 public:
   // Gets the state that the the blower system should (ideally) deliver right
   // now.
-  BlowerSystemState DesiredState(Time now, const VentParams &params) {
+  BlowerSystemState DesiredState(Time now, const VentParams &params,
+                                 const SensorReadings &readings) {
```

Comment:
Given that, per my comment above, it is kind of ambiguous whether updating internal state of the FSM should be done in desired_state() or in finished(), and that I'm finding it awkward that we're passing the same "now" to finished() and desired_state(), I wonder if it makes sense to instead put update into a separate function:

std::visit([]{fsm.update(now, readings)});
if (std::visit(...fsm.finished()...)) {...}
return std::visit(...fsm.desired_state()...);